src/stamps.ts

```diff
@@ -5,26 +5,46 @@ import { sleep } from './utils'
 const logger = console // eslint-disable-line no-console
 
 const DEFAULT_POLLING_FREQUENCY = 1_000
+const DEFAULT_STAMP_USABLE_TIMEOUT = 10_000
+
+interface Options {
+  pollingFrequency?: number
+  timeout?: number
+}
 
 /**
  * Wait until a given postage stamp is usable
  *
  * @param batchId Postage stamp id to be waited on
  * @param beeDebug Connection to the bee debug service
- * @param pollingFrequency (optional) How often should the stamp be checked in ms, default 1000
+ * @param options
+ *        pollingFrequency (optional) How often should the stamp be checked in ms, default 1000
+ *        timeout (optional) How long should the system wait for the stamp to be usable in ms, default to 10000
  */
 export async function waitUntilStampUsable(
   batchId: BatchId,
   beeDebug: BeeDebug,
-  pollingFrequency = DEFAULT_POLLING_FREQUENCY,
+  options: Options = {},
 ): Promise<DebugPostageBatch> | never {
-  // eslint-disable-next-line no-constant-condition
-  while (true) {
-    const stamp = await beeDebug.getPostageBatch(batchId)
+  const timeout = options.timeout || DEFAULT_STAMP_USABLE_TIMEOUT
+  const pollingFrequency = options.pollingFrequency || DEFAULT_POLLING_FREQUENCY
+
+  const timeoutPromise = async (): Promise<never> =>
+    new Promise((_, reject) =>
+      setTimeout(() => reject(new Error('Wait until stamp usable timeout has been reached')), timeout),
+    )
 
-    if (stamp.usable) return stamp
-    await sleep(pollingFrequency)
+  const stampWaitPromise = async (): Promise<DebugPostageBatch> | never => {
+    // eslint-disable-next-line no-constant-condition
+    while (true) {
```

Comment:
Hmmm, I am not sure if this is a good approach. Honestly, I think the `while` loop will then still run until the batch will be usable even the timeout will be reached. I would suggest using "cancellable token". Eq. not use `Promise.race` which works for users of this function, but does not clean up the resources (as Promises are not cancellable). I would suggest using the timeout to set a variable which then terminates the `while` loop and exits with a timeout error.

Reply:
This is good remark, implemented just that but I am throwing the error in the timeout directly as well otherwise with low refresh rates you could have a "longer timeout than expected". E.g `timeout=30_000` and `refreshRate=25_000` would "timeout" after 50 seconds and not 30 seconds.